[QOLDEV-905] add more deadlock detection to the datastore

ThrawnCA opened the pull request:
- Apply short lock timeouts anywhere the write-enabled engine is used

ThrawnCA committed:
[QOLDEV-905] add more deadlock detection to the datastore

- Apply short lock timeouts anywhere the write-enabled engine is used

duttonw reviewed (changes requested):
please also add to line 2127, resource_fields,
before the 
engine.execute(meta_sql)

ThrawnCA commented:
> please also add to line 2127, resource_fields, before the engine.execute(meta_sql)

That's problematic, since we're not dealing with a connection at that point, but the engine itself. Moreover, it's the read-only engine. It _shouldn't_ need the lock timeout, not if we've applied the timeout correctly to all the uses of the write-enabled engine.

I'll do some more digging.

ThrawnCA commented:
https://github.com/sqlalchemy/sqlalchemy/issues/5405 mentions that the direct `engine.execute` is deprecated, creating a connection object is preferred and might handle transactional behaviour better. I'll check if CKAN 2.11 has changed it.

duttonw commented:
interesting: https://docs.sqlalchemy.org/en/13/core/connections.html#understanding-autocommit

ThrawnCA committed:
[QOLDEV-905] clean up datastore handling to more closely match CKAN 2.11

- Simplify imports
- Use context manager to handle transaction cleanup

duttonw commented:
had another deadlock,
```SQL
1124: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
1400: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
3442: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
3991: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
6442: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
7504: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c" --blocking on  7485
...

7485: DROP INDEX "b8101b2c0004e5fa9b06b8893fdb0bff5190f71b" CASCADE --blocking on  7498

7498: SELECT count(_id) FROM "2bbef99e-9974-49b9-a316-57402b00609c"
```
killed session 7498 and it all cleared.

duttonw commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -2091,16 +2086,17 @@ def search_sql(self, context: Context, data_dict: dict[str, Any]):
-        resources_sql = sqlalchemy.text(
+        resources_sql = sa.text(
             u'''SELECT alias_of FROM "_table_metadata" WHERE name = :id''')
         results = self._get_read_engine().execute(resources_sql, id=alias)
```
it seems we have a mix of write engine and read engine and direct (missing the wrappers).

someparts are in the writing and others are in the reading. (And should not be mixed). since datastore/xloader can split reading and writing ops.

ThrawnCA committed:
[QOLDEV-905] add more deadlock detection to the datastore

- Apply short lock timeouts to looking up resource fields
- TODO investigate why read-only queries are blocking each other

ThrawnCA committed:
[QOLDEV-905] clean up type checking

ThrawnCA committed:
[QOLDEV-905] ensure lock timeout is applied to dropping datastore indexes

duttonw commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -700,11 +703,15 @@ def _drop_indexes(context: Context, data_dict: dict[str, Any],
+        indexes_to_drop = context['connection'].execute(
+            sql_get_index_string, data_dict['resource_id']).fetchall()
+
+    with context['connection'].begin():
```
``with`` now self closes (commits) the now defined transaction boundary instead of assuming that it is doing the correct thing.

duttonw commented on ckanext/datastore/backend/postgres.py:
```diff
@@ -2115,123 +2118,135 @@ def resource_fields(self, id: str) -> dict[str, Any]:
-                SELECT table_type FROM INFORMATION_SCHEMA.TABLES
-                WHERE table_name = '{0}'
+            with engine.connect() as conn:
+                conn.execute(_LOCK_TIMEOUT_SQL)
```
this should fix the lock we were seeing on ``SELECT count(_id) FROM "{0}`` which is good.
